applications/RANSApplication/python_scripts/formulations/turbulence_models/two_equation_turbulence_model_rans_formulation.py

```diff
@@ -0,0 +1,81 @@
+from KratosMultiphysics.RANSApplication.formulations.utilities import GetConvergenceInfo
+
+class TwoEquationTurbulenceModelRansFormulation(RansFormulation):
+    def __init__(self, model_part, settings, formulation_1, formulation_2):
+        super().__init__(model_part, settings)
+
+        self.stabilization_method = settings["stabilization_method"].GetString()
+
+        self.formulation_1 = formulation_1
+        self.formulation_1.SetStabilizationMethod(self.stabilization_method)
+        self.AddRansFormulation(self.formulation_1)
+
+        self.formulation_2 = formulation_2
+        self.formulation_2.SetStabilizationMethod(self.stabilization_method)
+        self.AddRansFormulation(self.formulation_2)
+
+        self.echo_level = settings["echo_level"].GetInt()
+        self.nu_t_convergence_utility = ScalarVariableDifferenceNormCalculationUtility(self.GetBaseModelPart(), Kratos.TURBULENT_VISCOSITY)
+        self.SetMaxCouplingIterations(settings["coupling_settings"]["max_iterations"].GetInt())
+
+    def GetMinimumBufferSize(self):
+        if (self.is_steady_simulation):
+            return 1
+        else:
+            return 2
+
+    def SetTimeSchemeSettings(self, settings):
+        if (settings.Has("scheme_type")):
+            scheme_type = settings["scheme_type"].GetString()
+            if (scheme_type == "steady"):
+                self.is_steady_simulation = True
+                self.GetBaseModelPart().ProcessInfo.SetValue(Kratos.BOSSAK_ALPHA, 0.0)
+            elif (scheme_type == "bdf2" or scheme_type == "bossak"):
+                self.is_steady_simulation = False
+                default_settings = Kratos.Parameters('''{
+                    "scheme_type": "PLEASE_SPECIFY_SCHEME_TYPE",
+                    "alpha_bossak": -0.3
+                }''')
+                settings.ValidateAndAssignDefaults(default_settings)
+                self.GetBaseModelPart().ProcessInfo.SetValue(Kratos.BOSSAK_ALPHA, settings["alpha_bossak"].GetDouble())
+            else:
+                raise Exception("Only \"steady\", \"bdf2\" and \"bossak\" scheme types supported. [ scheme_type = \"" + scheme_type  + "\" ]")
+        else:
+            raise Exception("\"scheme_type\" is missing in time scheme settings")
+
+        super().SetTimeSchemeSettings(settings)
+
+    def SolveCouplingStep(self):
+        settings = self.GetParameters()
+        relative_tolerance = settings["coupling_settings"]["relative_tolerance"].GetDouble()
+        absolute_tolerance = settings["coupling_settings"]["absolute_tolerance"].GetDouble()
+        max_iterations = self.GetMaxCouplingIterations()
+
+        for iteration in range(max_iterations):
+            self.nu_t_convergence_utility.InitializeCalculation()
+
+            self.ExecuteBeforeCouplingSolveStep()
+            for formulation in self.GetRansFormulationsList():
+                if (not formulation.SolveCouplingStep()):
+                    return False
```

Comment:
I don't fully understand this (please correct me if I'm wrong). For each RANS formulation you solve the coupling and if it returns `True` you return `False` in the outer `SolveCouplingStep`?

Reply:
I am sorry it is a copy paste mistake. I was using this safe guard in the base `RansFormulation`, but I don't need it here. So I removed it.

The situation is as following... (for the safe guard)
These `two_equation_turbulence_model_rans_formulation` does not know with which kind of V-P solver will be used with them. And, these `two_equation_turbulence_model_rans_formulation` should only be solved after V-P solve. So this is the safe guard I am using. It works like the following.

I always add the V-P `RansFormulation` first followed by `two_equation_turbulence_model_rans_formulation`. So `two_equation_turbulence_model_rans_formulation` always check whether V-P is solved, if only V-P is solved, then `two_equation_turbulence_model_rans_formulation` is solved, otherwise solving of `two_equation_turbulence_model_rans_formulation` will be skipped.